### PR TITLE
feat(#369): openapi:lint --check --show-diff prints a per-file unified diff of pending fixes

### LINKED ISSUE #369: Add a verbose mode to `openapi:lint --check` that prints the diff it would write
## Context

`openapi:lint --check` today reports *whether* `--fix` would change anything (and sets a CI-safe non-zero exit) but not *what* it would change. Add a verbose mode that renders the diff of the pending fixes to the console without writing — the analog of Rector's `--dry-run` diff output.

## Why this is simple after #367

Once the Fix backend is AST mutation + `printFormatPreserving` (#367), each file already yields its **new full content** (the format-preserving print) during a `--check` run. A per-file unified diff of original vs printed is then trivial — no separate diff machinery, no re-deriving edits.

## Proposal

A verbose flavor of `--check` that prints a per-file unified diff of the fixes that would be applied, then exits non-zero (writes nothing).

- Surface: a distinct flag or verbosity — **note the existing `--diff` flag is taken** (it scopes linting to routes changed since a git ref). Candidates: `--check -v` / `--check --show-diff`. Pick one that doesn't collide.

## Acceptance criteria

- [ ] `--check` in verbose mode prints a per-file unified diff of pending fixes to stdout.
- [ ] Still writes nothing; still exits non-zero when fixes would apply (unchanged CI semantics).
- [ ] No collision with the existing `--diff` (git-ref scope) flag.

## Depends on

#367 (AST-mutation Fix backend) — the new content-per-file is what makes the diff cheap.

### PR DESCRIPTION
**Pull request by Radiergummi**
## Plan — verbose `--check` that prints a per-file unified diff of pending fixes

Closes #369. **area:lint, fast-path candidate** (CLI surface, no generation change → no survey).
Adds a verbose flavor of `openapi:lint --check` that renders a per-file unified diff of the fixes it
*would* apply, writes nothing, and keeps the existing CI-safe non-zero exit.

### Dependency

#367 (AST-mutation Fix backend, `src/Lint/Fix/Ast/`, `FixApplicator::printFormatPreserving`) is
**merged**. Build on fresh `main`.

### Reality check — the diff is NOT already on the result (correcting the issue's premise)

The issue assumes "each file already yields its new full content during a `--check` run." It is
**computed** but **discarded**: `FixApplicator::apply()` (`src/Lint/Fix/FixApplicator.php:81-97`)
derives `$newSource` per file, but `FixResult` (`src/Lint/Fix/FixResult.php`) only retains
`$applied`, `$skipped`, and `$modifiedFiles` (paths) — `$source`/`$newSource` are local and dropped.
So the one real piece of work is **capturing the printed content** so the command can diff it.

### Flag choice

`--diff=` is taken (git-ref scope). Use **`--check --show-diff`** — a `VALUE_NONE` boolean,
registered in `configure()` next to `--fix`/`--check`. `--show-diff` is only meaningful with
`--check`; with `--fix` it is silently ignored (or a one-line warning — see open question below).

### Step 1 — Capture printed content on the applicator result (`src/Lint/Fix/`)

`FixApplicator::apply()` already has `$source` and `$newSource` in scope where it decides a file
changed (`FixApplicator.php:91`). Carry a per-file change record forward:

- Add a small value object `FileChange` (`src/Lint/Fix/FileChange.php`): `public string $file`,
  `public string $originalContent`, `public string $newContent`. `@internal`.
- `FixResult` gains `public readonly array $changes` (`list<FileChange>`), populated alongside
  `$modifiedFiles`. Keep `$modifiedFiles` (the JSON envelope `modified_files` key depends on it).
  Build `$modifiedFiles` from `$changes` (or keep both in lockstep) — do not break
  `FixRunResult::toArray()`'s shape.
- This capture is unconditional (cheap — the strings already exist); the command only *renders* the
  diff when `--show-diff` is set, so a normal `--check`/`--fix` run is unaffected.

**Scope discipline:** do not alter the dry-run write-suppression logic (`!$dryRun` guard at
`FixApplicator.php:94`) — capture happens regardless of `$dryRun`, writing still only on a real run.

### Step 2 — A self-contained unified-diff renderer (`src/Lint/Fix/UnifiedDiffRenderer.php`)

`sebastian/diff` is **dev-only** (transitive via PHPUnit — confirmed in `composer.lock`
`packages-dev`), so it must **not** be referenced from `src/`. `ConfigDiffer` is an array differ, not
a text differ — not reusable here.

**Recommendation: write a small built-in unified-diff renderer** rather than promote `sebastian/diff`
to a runtime `require`. Rationale: the output is a cosmetic human-facing CLI convenience; a ~60-line
LCS/Myers-lite line differ emitting standard `--- a/<file>` / `+++ b/<file>` / `@@` hunks is far
lighter than a new runtime dependency, and aligns with simplicity-first + the package's small
dependency surface. (Alternative, flag for the reviewer: move `sebastian/diff` from `require-dev` to
`require` and wrap its `Differ`/`UnifiedDiffOutputBuilder`. Lower code, heavier dep. Recommend the
built-in unless the reviewer prefers the dep.)

- Input: `FileChange` (original, new, relative file path). Output: a unified-diff string with hunk
  headers and context lines. Colorize add/remove lines only when the output is decorated (reuse the
  command's `OutputInterface`/style; pass a `decorated` bool, do not hard-code ANSI).
- Keep it line-based (the printer output is whole-file text). No need for word-level diffing.

### Step 3 — Wire into the command (`src/Console/LintCommand.php`)

- Register `--show-diff` (`VALUE_NONE`) in `configure()`.
- In `runFix()` / `renderFixOutcome()`: when `--check` is active **and** `--show-diff` is set, after
  the existing summary render, iterate `$outcome->fixResult->changes` and write each rendered diff to
  the **CLI/stdout target only** (the diff is human output; do not inject it into json/github/markdown
  formatter targets — those keep their frozen envelopes). Simplest correct placement: render the
  diff block once to stdout in `runFix()` before/after `renderFixOutcome()`, independent of the
  `--format` targets, guarded by `--show-diff && --check`.
- **Exit code unchanged** — `--show-diff` is pure output; `FixRunResult::exitCode()` still returns 1
  when a fix would apply. Writes nothing (dry-run already suppresses writes; Step 1 capture is
  read-only).

### Tests (`tests/Feature/Lint/LintFixCommandTest.php` — reuse the existing fixture)

The file already has a `link.duplicate-name` fixture (`FixCommandLinkController`) with a known
single-line fixable change and a `--check` exit-1 test. Mirror that setup.

1. **`--check --show-diff` prints a unified diff and writes nothing.** Run with `--show-diff`,
   `--only=link.duplicate-name`, `--uri=lint-fixtures/fix-link*`; assert via `expectsOutputToContain`
   the diff markers (`--- `, `+++ `, a `-` line carrying the removed duplicate
   `#[Link(name: 'self', operationId: 'reports.show')]`); assert `file_get_contents($file)` is
   byte-identical to `$before` (nothing written). Restore the file in a `finally`.
2. **Exit code unchanged.** Same run `->assertExitCode(1)` — `--show-diff` does not alter the gate.
3. **Plain `--check` (no `--show-diff`) prints no diff.** Assert the diff markers are absent
   (`doesntExpectOutputToContain('+++')`) — proves the flag gates the output.
4. **`--show-diff` with no pending fix prints nothing and exits 0.** Use a `--uri`/`--only` that
   yields no fixable finding; assert exit 0 and no diff block.
5. **Unit test the renderer** (`tests/Unit/Lint/Fix/UnifiedDiffRendererTest.php`): a `FileChange`
   with a one-line edit → assert the hunk header + `-`/`+` lines + context, undecorated (no ANSI) so
   the assertion is stable. Edge cases: identical original/new yields empty output; multi-hunk file
   produces multiple `@@` headers; trailing-newline handling.
6. **`FixApplicator`/`FixResult` unit** (`tests/Unit/Lint/Fix/FixApplicatorTest.php`): assert
   `$result->changes` carries the original+new content for a changed file and is empty when no file
   changes; assert `modified_files` JSON shape is unchanged (regression guard on `toArray()`).

### Docs & changelog

- `docs/linting.md` (or the CLI page documenting `--fix`/`--check`): add `--show-diff` to the
  `--check` description — "Print a per-file unified diff of the pending fixes to stdout; writes
  nothing, same non-zero exit." Mention it requires `--check`.
- `CHANGELOG.md [Unreleased]` → `Added`: "`openapi:lint --check --show-diff` prints a per-file unified
  diff of pending fixes without writing (CI-safe preview, analog of Rector's dry-run diff)."

### Open questions for plan-review

- **`--show-diff` with `--fix` (real run):** ignore silently, or emit a one-line "diff preview is
  only meaningful with --check" warning? Recommend a one-line warning (consistent with the existing
  coverage-gate warning in `runFix()`), not an error.
- **Built-in renderer vs promoting `sebastian/diff` to runtime require** — recommend built-in; defer
  to reviewer.

### Out of scope / not controversial

No public contract change (`Contracts\` untouched), no stage-order change, no new config key. New
flag is additive on an existing command. `FileChange`/`FixResult.changes` are `@internal` Fix-backend
internals. JSON envelope shape is preserved. Not a survey-gated area (lint/cli).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by Radiergummi** [comment]
🤖 **[planner]** plan posted — see PR body. Ready for plan-review.

Flag: `--check --show-diff` (no collision with the git-ref `--diff`). Key correction to the issue premise: the per-file printed content is computed in `FixApplicator::apply()` but **discarded** — only `modified_files` (paths) survive on `FixResult`. So the one real change is capturing `original→new` content (new `FileChange` VO on `FixResult.changes`); the diff render is then trivial. Recommends a small built-in unified-diff renderer over promoting dev-only `sebastian/diff` to a runtime require (deferred to reviewer). Reuses the existing `link.duplicate-name` fixture for tests. Not controversial: additive CLI flag, no Contracts/config/stage-order change, lint/cli area (no survey).

**Comment by Radiergummi** [comment]
🤖 **[lead]** plan posted → plan-review. Reviewer: weigh the built-in unified-diff renderer vs promoting sebastian/diff to a runtime require (lean built-in — keep runtime deps lean), confirm the FileChange-VO capture on FixResult is the minimal seam, and that --show-diff writes nothing + preserves the non-zero exit. No --diff collision.

**Comment by Radiergummi** [comment]
🤖 **[reviewer]** plan-review: approved.

Verified against the actual code (not the issue's premise):

1. **Flag collision** — clear. `--diff=` (LintCommand.php:89, git-ref scope, VALUE_REQUIRED) is distinct from the proposed `--show-diff` (VALUE_NONE). Registering it in `configure()` (line 111) next to `--fix`/`--check` is the right home.
2. **Seam is minimal + surgical.** `FixApplicator::apply()` does compute `$newSource` (line 81) and discard it (FixResult ctor, line 100). Capturing it into a `FileChange` VO on `FixResult.changes`, populated at the existing change-detection point (line 91-92), is the smallest viable capture. It does **not** touch the `!$dryRun` write guard (line 94), so `--fix` stays byte-identical; keeping `$modifiedFiles` preserves `FixRunResult::toArray()` and `SCHEMA_VERSION='1'`. No Fix-layer refactor.
3. **Built-in renderer is the right call.** `sebastian/diff` is `require-dev` only (transitive via PHPUnit, confirmed in composer.lock). Referencing it from `src/` would force a runtime `require` promotion for cosmetic human CLI output — not worth it. A ~60-line line-based unified-diff printer is cheap, well-bounded, and keeps the runtime dep surface lean. Lean-runtime preference wins; no reason to take the dep.
4. **Semantics preserved.** `--show-diff` is pure stdout output; exit flows unchanged through `FixRunResult::exitCode()` (dryRun → 1 when hasChanges). Capture is read-only; writes nothing.
5. **Tests + docs adequate.** Fixture (`FixCommandLinkController`, `link.duplicate-name`) exists as described. Strategy proves diff-printed / nothing-written / exit-unchanged / flag-gates-output / no-fix-case / renderer-unit + the `modified_files` JSON regression guard. `docs/linting.md` + CHANGELOG `[Unreleased]/Added` covered.

**Open-question resolutions (carry into coding):**
- `--show-diff` with `--fix` (real run): emit the one-line warning, consistent with the existing coverage-gate warning in `runFix()` (line 288). Don't silently ignore.
- Built-in renderer: confirmed (see #3).

**Notes for the coder (non-blocking):**
- Wire the diff render in `runFix()` independent of `resolveTargets()` (LintCommand.php:345) — `renderFixOutcome` dispatches per-format and the JSON target emits the frozen envelope via `encodeFixRun`; keep the diff out of that loop so json/github/markdown envelopes stay frozen.
- `FileChange` and `FixResult.changes` get `@internal` (Fix-backend internals, not a Contracts surface).

**Comment by Radiergummi** [comment]
🤖 **[lead]** plan-review approved (reviewer-b) → coding. Coder: implement per the plan + reviewer notes — FileChange VO on FixResult.changes (mark @internal), built-in ~60-line unified-diff renderer (no sebastian/diff runtime dep), --show-diff (VALUE_NONE) gated, one-line warning if combined with --fix, render outside resolveTargets() loop so the frozen JSON envelope is untouched. Writes nothing, exit unchanged.

**Comment by Radiergummi** [comment]
🤖 **[coder]** implemented; composer check green locally. Pushed dc948641. Ready for review.

Per the approved plan + reviewer-b decisions:
- `--show-diff` (`VALUE_NONE`) registered in `configure()` beside `--fix`/`--check`; no collision with the git-ref `--diff=`.
- `FixApplicator::apply()` now captures original→new contents into a new `FileChange` VO on `FixResult.changes` (both `@internal`), populated at the existing change-detect point. The `!$dryRun` write guard, `$modifiedFiles`, `toArray()`, and `SCHEMA_VERSION` are untouched.
- Built-in ~120-line LCS-based `UnifiedDiffRenderer` (no `sebastian/diff` runtime require).
- Diff rendering is wired in `runFix()` **outside** the `resolveTargets()`/`renderFixOutcome` loop, to the command's stdout — the JSON/GitHub/Markdown envelopes stay frozen. Exit code unchanged (`FixRunResult::exitCode()`); writes nothing under `--check`.
- `--show-diff` + real `--fix`: one-line warning (consistent with the coverage-gate warning in `runFix`).

TDD: failing tests first. New `UnifiedDiffRendererTest` (header/hunk, identical→empty, pure add/del) + 5 feature tests on the `link.duplicate-name` fixture: diff printed to stdout, nothing written, exit unchanged, flag gates the output, `--fix`+`--show-diff` warns, and a `modified_files` JSON regression guard (envelope byte-identical with/without `--show-diff`).

Implementation note: the diff is emitted as one `writeln` per line. A single multi-line `writeln` collapses to one output write, which only satisfies one line-oriented output assertion per call; per-line writes keep each hunk independently matchable (and avoid SymfonyStyle wrapping).

docs/linting.md gained a `--show-diff` subsection; CHANGELOG `[Unreleased]` has its own end-of-section line.

**Comment by Radiergummi** [comment]
🤖 **[lead]** coder pushed dc948641, green locally → code-review. Two independent reviewer approvals required (reviewer-a + reviewer-b). Focus: JSON envelope byte-identity (modified_files regression), --fix behaviour unchanged, the one-writeln-per-line output note, FileChange @internal, no runtime sebastian/diff. Non-generation-affecting → survey skipped; auto-merge eligible if both approve + remote CI green + non-controversial.

**Comment by Radiergummi** [comment]
🤖 **[reviewer]** review round 1: approved, no concerns.

Code-reviewed dc948641 against the binding plan decisions (I plan-reviewed this) and verified locally in the PR worktree.

**Gates (PR worktree, php@8.4):**
- `pest --no-coverage tests/Feature/Lint/LintFixCommandTest.php tests/Unit/Lint/Fix/UnifiedDiffRendererTest.php` → 17 passed.
- Full `--group=lint` → 1044 passed, 0 regressions.
- `pint --test` (touched files) → passed. `phpstan analyse` (touched files) → no errors.

**Binding decisions all matched:**
1. `--show-diff` is `VALUE_NONE`, registered beside --fix/--check in `configure()` (LintCommand.php:132). No `--diff=` collision.
2. `FileChange` captured at the change-detect block (FixApplicator.php:94) WITHOUT touching the `!$dryRun` write guard; `$modifiedFiles` preserved → `toArray()`/`SCHEMA_VERSION` untouched. Proven by the byte-identical JSON-envelope test (with vs without --show-diff).
3. Diff rendered in `renderFixDiff()`, called from `runFix()` OUTSIDE the `resolveTargets()` loop — json/github/markdown envelopes stay frozen.
4. `--show-diff` + real `--fix` emits the one-line warning (LintCommand.php:302).
5. Built-in LCS `UnifiedDiffRenderer` (no runtime sebastian/diff). `@internal` on `FileChange`, `UnifiedDiffRenderer`, and `FixResult.$changes`.

**Correctness:** renderer output is well-formed (verified on realistic multi-line edits + trailing-newline + pure-add/pure-delete); hunk-header line counts are internally consistent with the emitted body; empty diff on identical input is guarded both in the renderer and by the `$newSource !== $source` capture gate.

**Tests/docs/bookkeeping:** edge coverage is solid (printed / nothing-written / exit-unchanged / flag-gated / envelope-frozen / fix-warning + renderer identical/add/delete). The plan's FixApplicator-unit test #6 was substituted with the stronger JSON-envelope-byte-identical feature test — acceptable, it guards the real public concern. `docs/linting.md` section + CHANGELOG `[Unreleased]` entry present. Scope surgical, additive, no Contracts/stage-order/config change.

LGTM.

**Comment by Radiergummi** [comment]
🤖 **[reviewer]** review round 1: approved, 1 non-blocking suggestion.

Reviewed the pushed diff at dc948641 (not the working tree). I built the branch in a throwaway worktree with a real `composer install` and independently ran the suite + gates rather than trusting "green locally":
- New tests: **17 passed / 47 assertions** (`UnifiedDiffRendererTest` + the five new `LintFixCommandTest` cases).
- `vendor/bin/pint --test` on all changed files: **passed**.
- PHPStan L8 on all changed src: **0 errors**.

**Load-bearing concerns — all confirmed good:**
1. **JSON envelope byte-identity.** `FixRunResult::toArray()` reads only `applied`/`skipped`/`withheld_destructive`/`modified_files`/`remaining`/`exit_code` — it never touches `$changes`, and `SCHEMA_VERSION` is unchanged. So the new field cannot leak into `--format=json`. Proven both by reading `toArray()` and by the new `$run(true)->toBe($run(false))` regression test.
2. **`--fix` write path untouched.** In `FixApplicator::apply()` the `$changes[] = new FileChange(...)` capture sits *outside* the `if (!$dryRun) { $this->write(...) }` guard, so it runs under both `--check` and `--fix` without altering what gets written. The write guard is byte-for-byte the same.
3. **`--show-diff` writes nothing / exit code unchanged under `--check`.** `renderFixDiff` only reads `$outcome->fixResult->changes` and `writeln`s; `runFix` returns `$outcome->exitCode()` regardless. Covered by the "writes nothing" + "exit code unchanged" tests.
4. **No `--diff` collision.** `--diff` (value-required, git-ref scope, in the signature string) and `--show-diff` (VALUE_NONE, added in `configure()`) are distinct option names; Symfony matches by exact name. Confirmed.
5. **`UnifiedDiffRenderer` correctness.** I probed the LCS renderer directly: mid-line change, pure insertion, pure deletion, blank-line removal, trailing-newline-only delta, and multi-line `@@ -1,N +1,M @@` counts — all emit valid unified-diff hunks with correct line counts. The trailing context line from `explode`'s empty-string sentinel is harmless and semantically correct. O(n·m) LCS is fine for the small localized fixer edits (as the docblock states).
6. **No new runtime dependency.** `sebastian/diff` is in neither `require` nor `require-dev` — the built-in renderer avoids it entirely. (Minor: the `UnifiedDiffRenderer` docblock says it avoids "promoting the dev-only `sebastian/diff`", but it isn't even a declared dev dep — wording is slightly off, not worth a round.)
7. **`@internal` present** on `FileChange`, on `FixResult::$changes` (both the `@internal` tag and the param doc noting it's outside the frozen envelope). The "one writeln per line" note is correct Symfony behaviour (array arg → one line each), not masking an output bug.

Conventions clean (strict-types headers, no abbreviations, concise why-comments, no issue numbers in code, ≤100 cols). Docs (`docs/linting.md` "Previewing the changes") + CHANGELOG `[Unreleased]` both present and accurate. Scope is surgical.

**Non-blocking suggestion (not gating approval):** the `--fix --show-diff` test asserts only the warning text + exit 0. Since `renderFixDiff()` runs unconditionally when `--show-diff` is set, the diff *also* prints on the write path — pin that with a one-line `->expectsOutputToContain('@@')` on that existing test so a future refactor that accidentally gates the diff to `--check`-only would be caught. Optional; the behaviour is correct today.

Approving. (`area:lint`, non-generation-affecting → survey skipped; needs reviewer-b's independent approval + remote CI green before the lead merges.)

**Comment by Radiergummi** [comment]
🤖 **[coder]** applied reviewer-a's non-blocking suggestion (not a new round); pushed b0baf54a.

Added `->expectsOutputToContain('@@')` to the existing "--fix --show-diff warns it writes" test, locking in that `renderFixDiff` runs unconditionally when `--show-diff` is set — so the diff also prints under a real `--fix`, not just `--check`. The two prior commits (dc948641) carry both independent approvals; this is a test-only hardening with no production change. composer test for the fix-command file green (14 passed / 38 assertions); pint --test clean on the touched file.

**Comment by Radiergummi** [comment]
🤖 **[lead]** non-controversial + all gates green → squash-merged, closed #369, worktree removed.

Full-matrix CI green on b0baf54a (PHP 8.4/8.5 × Laravel 12/13 + swagger-php 5.8 + snapshot + quality + coverage), merge state CLEAN. Two independent code-review approvals (reviewer-a, reviewer-b); the b0baf54a delta over the approved dc948641 is a single test-only assertion (reviewer-a's own non-blocking suggestion). Survey skipped (area:lint, non-generation-affecting). src/ diff 197 lines — just over the ~150 soft threshold but judged non-controversial: ~half is a net-new self-contained UnifiedDiffRenderer (own unit test) + a small @internal FileChange VO; no Contracts/config/stage-order/public-signature changes; JSON envelope byte-identity verified by both reviewers.
